### PR TITLE
Fix quota controller discovery deadlock

### LINKED ISSUE #71450: ResourceQuota objects not refreshing anymore (while admission controller still working)
**What happened**

ResourceQuotas are configured in every namespace of a busy production cluster. At some point the following happened:

- "status:used" got inflated / was more than the sum of Requests in namespaces
- "status:hard" refused to get updated via applying new manifests - "spec:hard" was normally updated but "status:hard" was stuck.
- attempted deleting and recreating the ResourceQuota, and "status" remained empty {}.

Still, the main functionality of refusing to schedule new pods that exceeded current status:hard was still working.

**Remediation:**

killed controller-manager, as soon as another pod took the master lease and controllers started working, the issue got resolved.

My best guess is that there's some kind of deadlock in the ResourceQuota controller or in controller-manager in general.

**Environment**:
- k8s 1.10.9
- Cloud provider: VSphere on-premise
- OS: Ubuntu 16.04
- Kernel:  4.4.0-135

/kind bug
/sig apps api-machinery scheduling
(not sure what's the exact SIG for this)

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

Fixes a discovery deadlock that can occur during quota controller discovery sync.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #71450

**Special notes for your reviewer**:

Similar fix to https://github.com/kubernetes/kubernetes/pull/64235

**Does this PR introduce a user-facing change?**:
```release-note
Fixes potential deadlock in resource quota controller
```

### CODE REVIEW

**Comment by liggitt** [comment]
/hold
failures look related (and very interesting... persistent `resource_quota_controller.go:460] failed to sync resource monitors (attempt 379): couldn't start monitor for resource "extensions/v1beta1, Resource=networkpolicies": unable to monitor quota for resource "extensions/v1beta1, Resource=networkpolicies"` error)

File: pkg/controller/resourcequota/resource_quota_controller.go [hunk 1]
```diff
@@ -423,24 +423,90 @@ func (rq *ResourceQuotaController) Sync(discoveryFunc NamespacedResourcesFunc, p
...
+			// On a reattempt, check if available resources have changed
+			if attempt > 1 {
+				// Get the current resource list from discovery.
+				newResources, err = GetQuotableResources(discoveryFunc)
+				if err != nil {
+					utilruntime.HandleError(err)
+					if discovery.IsGroupDiscoveryFailedError(err) && len(newResources) > 0 {
+						// In partial discovery cases, don't remove any existing informers, just add new ones
+						for k, v := range oldResources {
+							newResources[k] = v
```
Comment (logicalhan) [inline pkg/controller/resourcequota/resource_quota_controller.go]: Is this correct? This line seems to overwrite values but the comment seems to imply we only want to add values if the keys are non-existing.

Comment (logicalhan) [reply pkg/controller/resourcequota/resource_quota_controller.go]: Sorry, please ignore, I realized later that this was basically just a set thing.

**Comment by liggitt** [comment]
this isn't relevant until we actually use dynamic discovery for quota

/close

**Pull request closed without merging**